src/Functions/ReplaceRegexpImpl.h

```diff
@@ -48,45 +51,75 @@ struct ReplaceRegexpImpl
 
     static constexpr int max_captures = 10;
 
+    /// The replacement string references must not contain non-existing capturing groups.
+    static void checkSubstitutions(std::string_view replacement, int num_captures)
+    {
+        for (size_t i = 0; i < replacement.size(); ++i)
+        {
+            if (replacement[i] == '\\' && i + 1 < replacement.size())
+            {
+                if (isNumericASCII(replacement[i + 1])) /// substitution
+                {
+                    int substitution_num = replacement[i + 1] - '0';
+                    if (substitution_num >= num_captures)
+                        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Substitution '\\{}' in replacement argument is invalid, regexp has only {} capturing groups", substitution_num, num_captures - 1);
+                }
+            }
+        }
+    }
+
     static Instructions createInstructions(std::string_view replacement, int num_captures)
     {
+        checkSubstitutions(replacement, num_captures);
+
         Instructions instructions;
 
         String literals;
+        literals.reserve(replacement.size());
+
         for (size_t i = 0; i < replacement.size(); ++i)
         {
             if (replacement[i] == '\\' && i + 1 < replacement.size())
             {
-                if (isNumericASCII(replacement[i + 1])) /// Substitution
+                if (isNumericASCII(replacement[i + 1])) /// substitution
                 {
                     if (!literals.empty())
                     {
                         instructions.emplace_back(literals);
                         literals = "";
                     }
-                    instructions.emplace_back(replacement[i + 1] - '0');
+                    int substitution_num = replacement[i + 1] - '0';
+                    instructions.emplace_back(substitution_num);
                 }
                 else
-                    literals += replacement[i + 1]; /// Escaping
+                    literals += replacement[i + 1]; /// escaping
                 ++i;
             }
             else
-                literals += replacement[i]; /// Plain character
+                literals += replacement[i]; /// plain character
         }
 
         if (!literals.empty())
             instructions.emplace_back(literals);
 
-        for (const auto & instr : instructions)
-            if (instr.substitution_num >= num_captures)
-                throw Exception(
-                    ErrorCodes::BAD_ARGUMENTS,
-                    "Id {} in replacement string is an invalid substitution, regexp has only {} capturing groups",
-                    instr.substitution_num, num_captures - 1);
-
         return instructions;
     }
 
+    static bool canFallbackToStringReplacement(const String & needle, const String & replacement, const re2::RE2 & searcher, int num_captures)
+    {
+        if (searcher.NumberOfCapturingGroups())
```

Comment:
Could you explain, what this if checks, please?

Reply:
Function `canFallbackToStringReplacement` checks if the pattern is "trivial" in which case we use fast plain string replacement. A "trivial" pattern means
- the regular expression has no metacharacters (e.g. `*`, `{`, etc.). This is checked by l. 115-120).
- the regular expression has no capturing groups (see [here](https://github.com/google/re2/wiki/Syntax) and [here](https://clickhouse.com/docs/en/sql-reference/functions/string-replace-functions#replaceregexpone), basically `(` and `)`). This is checked by l. 110.

One might argue that the latter condition should be checked already by the former (`OptimizedRegularExpression::analyze`, 119). The thing is that `::analyze` is not free (+ I don't trust it 100% TBH, it is quite complex and written by ourselves, i.e. not part of re2). When `canFallbackToStringReplacement` is called, `searcher.NumberOfCapturingGroups()` is free as the re2 regexp `searcher` was already compiled at this point. Hence, we can call it to save a little time, potentially.